third_party/intel/lib/Target/LLVMIR/DSE.cpp

```diff
@@ -0,0 +1,426 @@
+      //  the same address, yielding stale cache entries.
+      if (IsMemDef && DeadInst->getType()->isVoidTy())
+        DeadInst->eraseFromParent();
+      else
+        ToRemove.push_back(DeadInst);
+    }
+  }
+
+  Function &F;
+  AliasAnalysis &AA;
+  MemorySSA &MSSA;
+  DominatorTree &DT;
+  const TargetLibraryInfo &TLI;
+  const LoopInfo &LI;
+
+  // All MemoryDefs that potentially could kill other MemDefs.
+  SmallVector<MemoryDef *, 64> MemDefs;
+  // Any that should be skipped as they are already deleted
+  SmallPtrSet<MemoryAccess *, 4> SkipStores;
+  // Keep track of all of the objects that are invisible to the caller after
+  // the function returns.
+  DenseMap<const Value *, bool> InvisibleToCallerAfterRet;
+  // Post-order numbers for each basic block. Used to figure out if memory
+  // accesses are executed before another access.
+  DenseMap<BasicBlock *, unsigned> PostOrderNumbers;
+  /// Keep track of instructions (partly) overlapping with killing MemoryDefs
+  /// per basic block.
+  MapVector<BasicBlock *, InstOverlapIntervalsTy> IOLs;
+  /// Dead instructions to be removed at the end of DSE.
+  SmallVector<Instruction *> ToRemove;
+};
+
+class SetAddressPayloadInfo {
+  friend raw_ostream &operator<<(raw_ostream &, const SetAddressPayloadInfo &);
+
+public:
+  static constexpr StringLiteral prefix =
+      "__builtin_IB_subgroup_setBlock2DAddressPayloadBlock";
+
+  /// The field to set in the address payload.
+  enum class Field { X, Y, Unknown };
+
+  SetAddressPayloadInfo() = default;
+
+  SetAddressPayloadInfo(CallInst *CI, MemoryAccess *memAccess)
+      : ptr(nullptr), val(nullptr), field(Field::Unknown),
+        memAccess(memAccess) {
+    const StringRef funcName = CI->getCalledFunction()->getName();
+    assert(funcName.starts_with(prefix) &&
+           "Expecting address payload set function");
+    assert(CI->arg_size() == 2 && "Expecting 2 arguments");
+
+    ptr = CI->getArgOperand(0);
+    val = CI->getArgOperand(1);
+    field = funcName.ends_with("X")
+                ? Field::X
+                : (funcName.ends_with("Y") ? Field::Y : Field::Unknown);
+    assert(field != Field::Unknown && "Expecting X or Y");
+  }
+
+  bool samePtrAndValueAndField(const SetAddressPayloadInfo &other) const {
+    return ptr == other.ptr && val == other.val && field == other.field;
+  }
+  bool samePtrAndField(const SetAddressPayloadInfo &other) const {
+    return ptr == other.ptr && field == other.field;
+  }
+  bool operator==(const SetAddressPayloadInfo &other) const {
+    return ptr == other.ptr && val == other.val && field == other.field &&
+           memAccess == other.memAccess;
+  }
+  bool operator<(const SetAddressPayloadInfo &other) const {
+    return &ptr < &other.ptr || (ptr == other.ptr && &val < &other.val) ||
+           (ptr == other.ptr && val == other.val && &field < &other.field) ||
+           (ptr == other.ptr && val == other.val && field == other.field &&
+            memAccess < other.memAccess);
```

Comment:
I think something like:
```suggestion
  bool operator<(const SetAddressPayloadInfo &other) const {
    if (ptr < other.ptr)
      return true;
    if (ptr > other.ptr)
      return false;
    if (val < other.val)
      return true;
    if (val > other.val)
      return false;
    if (static_cast<std::underlying_type_t<Field>>(field) < static_cast<std::underlying_type_t<Field>>(other.field))
      return true;
    if (static_cast<std::underlying_type_t<Field>>(field) > static_cast<std::underlying_type_t<Field>>(other.field))
      return false;
    ...
```
would be better. Note `&` does not compare values, but addresses and might yield weird results. Also note we are comparing pointers here, so order semantics might not be what we ideally want. Are we even using this? Cause if we are not, we may very well remove it.